src/main/java/org/opensearch/security/filter/SecurityFilter.java

```diff
@@ -378,7 +379,15 @@ private <Request extends ActionRequest, Response extends ActionResponse> void ap
                 log.trace("Evaluate permissions for user: {}", user.getName());
             }
 
-            final PrivilegesEvaluatorResponse pres = eval.evaluate(user, action, request, task, injectedRoles);
+            PrivilegesEvaluationContext context = null;
+            PrivilegesEvaluatorResponse pres;
+
+            try {
+                context = eval.createContext(user, action, request, task, injectedRoles);
+                pres = eval.evaluate(context);
+            } catch (PrivilegesEvaluatorResponse.NotAllowedException e) {
```

Comment:
I think this is going to get confusing over time if the only place NotAllowedException is thrown is within `createContext`, but not within `eval.evaluate`.

Reply:
wdyt about moving the cross-cluster specific validation to here instead of within `createContext`?

Reply:
I agree - having an exception here was certainly not my first choice. However, due to the validation performed the injected roles done at https://github.com/opensearch-project/security/blob/2b5a811de599f7c7fc2ca2b9246e57fd6cfaf33b/src/main/java/org/opensearch/security/privileges/PrivilegesEvaluator.java#L262 I needed a dual return type. 

The alternative would be not using an exception but a kind of `AlternativeResult<PrivilegesEvaluationContext, PrivilegesEvaluatorResponse>` type, but that would be just a different inconsistency. WDYT?

Reply:
I would be ok with `Pair<PrivilegesEvaluationContext, Optional<PrivilegesEvaluatorResponse>>`.

Reply:
Using a `Pair` type would lose the semantics that the results are alternatives, though.

Reply:
> Using a Pair type would lose the semantics that the results are alternatives, though.

`AlernativeResults` sounds good.

Reply:
> wdyt about moving the cross-cluster specific validation to here instead of within createContext?

You mean moving it to `eval.evaluate()`?

The problem is that validation and determination of roles is entangled.

 I'd like to keep `PrivilegesEvaluationContext` immutable - with caches as an exception. That means that the `mappedRoles` should be determined as soon as it is created.

But, maybe, I should give up the idea that the mapped roles should not change in the `PrivilegesEvaluationContext`  object. :thinking:

Reply:
> You mean moving it to eval.evaluate()? 

Potentially a method called `computeMappedRoles` that's called before createContext that can optionally return a PrivilegesEvaluatorResponse?

Reply:
> I would be ok with Pair<PrivilegesEvaluationContext, Optional<PrivilegesEvaluatorResponse>>. Curious to know what other maintainers think too.

Will we have different contexts at any given point? What are the pros and cons of having a catch block vs having a pair of AlternativeResult?

Reply:
> What are the pros and cons of having a catch block vs having a pair of AlternativeResult?

IMHO, it's mostly just a matter of taste.

Pro exception and catch:

- This is the "official" way in Java to signal exceptional returns from a method

Contra exception and catch:

- It might be confusing that not all methods throw exceptions but return `PrivilegeEvaluatorResponse` instead
- Exceptions have a slight performance penalty - even though it should be not an issue here, as we are looking at a rarely happening case

Pro AlternativeResult:

- Less confusing? (Well, I am not sure about this one)
- Bit better performance in the error case

Contra AlternativeResult:

- Requires more complicated logic to model the alternative cases.
- Correct usage cannot be verified statically during compile time. Instead, it needs to throw exceptions in case certain methods are used in the wrong state.

Reply:
@nibix wdyt about renaming the exception? I think it would be confusing to add an exception called NotAllowedException and many people may assume by looking at the code that a NotAllowedException is thrown when authz fails and that's not true.

Reply:
If we introduce the catch block then it will be only for this particular case, but rest of the places would still return PrivilegeEvaluatorResponse. It might be confusing to someone who reads it in future to understand why is there a catch block only at one place

Reply:
After having slept a night over it, my opinion is: Even though I'd like the `PrivilegeEvaluationContext` object as immutable as possible, I'd make the `mappedRoles` modifiable - that way we can move the validation code again to the actual `evaluate()` method and get a clean method interface instead. WDYT?

Reply:
I think it makes sense to be immutable. Did you see the comment [here](https://github.com/opensearch-project/security/pull/4490#discussion_r1661365425)? What about a new method that's called before `createContext` where the responsibility of the method is to compute the mapped roles?

Reply:
I prefer to not have the one off exception type here. The rest, is viable in either way but I would rather have limited exception types.

Reply:
@cwperks 

> I think it makes sense to be immutable. Did you see the comment https://github.com/opensearch-project/security/pull/4490#discussion_r1661365425? What about a new method that's called before createContext where the responsibility of the method is to compute the mapped roles?

But if I understand it correctly, that method also would need to validate the roles with the             ConfigConstants.OPENDISTRO_SECURITY_INJECTED_ROLES_VALIDATION value. Thus that method also would need a dual return type: either the mapped roles or a privilege evaluation error, either in form of an exception or another form of alternative return type. That would not really win anything.

Reply:
So, I now went ahead and moved the logic related to `OPENDISTRO_SECURITY_INJECTED_ROLES_VALIDATION` back to the evaluate() method. This allowed me to delete the exception stuff again, but at the cost of having the `mappedRoles` attribute non-final in `PrivilegesEvaluationContext`. 

Side note: All this is only necessary because of the call to `mappedRoles = ImmutableSet.copyOf(injectedRolesValidationSet)` in line 268 of the following code block:

https://github.com/opensearch-project/security/blob/20c87750524481242199ec8d63b41db56cd6e64f/src/main/java/org/opensearch/security/privileges/PrivilegesEvaluator.java#L255-L269

I have ready through the design docs which introduced the functionality and could not find a good reason why the roles need to be modified by a validation string. So, if we would determine that this is actually not necessary, we could make the attribute final again.

Reply:
Should we capture the role re-assignment on the validation set in an issue to discuss further? It looks safe to remove for me since it enforces that the roles must be a superset of the validation set.

I would try removing the line and see if there's an existing test that fails.